File: CodingInterviewsPythonSolutions/973. K Closest Points to Origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        maxHeap = []
        for idx in range(len(points)):
            loc = [points[idx][0], points[idx][1]]
            dist = loc[0] * loc[0] + loc[1] * loc[1]
            
            if len(maxHeap) < k or dist < -maxHeap[0][0]:
                heapq.heappush(maxHeap, (-dist, loc))
            
            if len(maxHeap) > k:
                heapq.heappop(maxHeap)
        
        result = []
        while maxHeap:
            result.append(heapq.heappop(maxHeap)[1])
        
        return result

File: CodingInterviewsJavaSolutions/973. K Closest Points to Origin.java
import java.util.PriorityQueue;

class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> getDistance(b) - getDistance(a));
        
        for (int[] point : points) {
            maxHeap.offer(point);
            
            if (maxHeap.size() > k)
                maxHeap.poll();
        }
        
        int[][] result = new int[k][2];
        while (k > 0) {
            result[--k] = maxHeap.poll();
        }
        
        return result;
    }
    
    private int getDistance(int[] point) {
        return point[0] * point[0] + point[1] * point[1];
    }
}
